janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/ArrayUtil.java

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.util;
+
+public class ArrayUtil {
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    /**
+     * Given an old capacity of an array, generate the new capacity that is larger than or equal to minimum required capacity.
+     * Under most circumstances, it doubles the old capacity. If the old capacity is already larger than half of @{link #MAX_ARRAY_SIZE},
+     * then new capacity will be {@link #MAX_ARRAY_SIZE}, which is slightly smaller than Integer.MAX_VALUE to avoid
+     * Out of Memory error. If the min capacity is smaller than old capacity, then old capacity will be simply returned.
```

Comment:
According to https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8206403:

> The number of bytes used for a VM's header of an array might vary by implementation but 8 seems to be a safe value, i.e., general upper bound.

Reply:
I strongly suspect 8 is safe, because it is used in the OpenJDK implementation of ArrayList as well:
https://hg.openjdk.java.net/jdk7/jdk7/jdk/file/tip/src/share/classes/java/util/ArrayList.java#l196

Reply:
Also, the implementation of `growSpace` itself is pretty close to what an `ArrayList` does (check link above). Thus I think this behavior has a solid foundation and is already widely used :+1:

Reply:
Thank you for bringing it here! Makes sense.